jax/_src/dtypes.py

```diff
@@ -249,6 +249,7 @@ def issubdtype(a, b):
 issubsctype = np.issubsctype
 
 # Enumeration of all valid JAX types in order.
+_python_scalar_types = [bool, int, float, complex]
 _weak_types = [int, float, complex]
```

Comment:
When do you need to separate between weak and scalar types? Like what's special about bool compared to other weak types?

Reply:
What's special about `bool` is that there is only a single `bool` width, and so there's no need to separate a strong bool from a weak bool in the context of type promotion. But this bug brought to light that we *do* need to treat boolean Python scalars as weakly-typed if we want type promotion to be invariant to JIT

Reply:
Consider adding this in a comment! Or not, in order to confuse our enemies

Reply:
@jakevdp Why not make it simpler? Would adding weak bool significantly affect performance?
It seems that originally "weak type" means "bitwidth unspecified", as they named `i_, f_, c_", but currently weak type changes type promotion behavior? As a result, single bitwidth doesn't mean there is no need to separate a weak bool from strong bool?

Reply:
Thanks for the comment. I ended up going the other direction, and ensuring that bools are treated as strong everywhere. It's a simpler change that keeps things consistent.